ProxySQL Crashed

Last night, one of our proxy instances crashed unexpected.

Attached the Logs.

Installed Version is 2.1.1 from the ProxySQL Debian Repo on Debian 10.8

/var/lib/proxysql/proxysql.log:
```
2021-06-30 05:38:06 main.cpp:1600:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2021-06-30 05:38:06 [INFO] ProxySQL version 2.1.1-40-g1c2b7e4
```

GDB output:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ff473c5a535 in __GI_abort () at abort.c:79
#2  0x00007ff473c5a40f in __assert_fail_base (fmt=0x7ff473dbcee0 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x55e2374ce404 "mybe", file=0x55e237474b1c "MySQL_Session.cpp", line=4217, function=<optimized out>) at assert.c:92
#3  0x00007ff473c68102 in __GI___assert_fail (assertion=assertion@entry=0x55e2374ce404 "mybe", file=file@entry=0x55e237474b1c "MySQL_Session.cpp", line=line@entry=4217, function=function@entry=0x55e237475810 <MySQL_Session::handler()::__PRETTY_FUNCTION__> "int MySQL_Session::handler()") at assert.c:101
#4  0x000055e236f1d728 in MySQL_Session::handler (this=this@entry=0x7ff46a46e800) at MySQL_Session.cpp:4506
#5  0x000055e236efa5ac in MySQL_Thread::process_all_sessions (this=this@entry=0x7ff46a400000) at MySQL_Thread.cpp:4890
#6  0x000055e236f01d3a in MySQL_Thread::run (this=this@entry=0x7ff46a400000) at MySQL_Thread.cpp:4424
#7  0x000055e236e99cec in mysql_worker_thread_func (arg=0x7ff4732bbeb0) at main.cpp:776
#8  0x00007ff474350fa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#9  0x00007ff473d314cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

--- a/lib/mysql_data_stream.cpp
+++ b/lib/mysql_data_stream.cpp
@@ -1352,9 +1352,16 @@ void MySQL_Data_Stream::return_MySQL_Connection_To_Pool() {
 	unsigned long long intv = mysql_thread___connection_max_age_ms;
 	intv *= 1000;
 	if (
-		( (intv) && (mc->last_time_used > mc->creation_time + intv) )
+		(( (intv) && (mc->last_time_used > mc->creation_time + intv) )
 		||
-		( mc->local_stmts->get_num_backend_stmts() > (unsigned int)GloMTH->variables.max_stmts_per_connection )
+		( mc->local_stmts->get_num_backend_stmts() > (unsigned int)GloMTH->variables.max_stmts_per_connection ))
+		&&
+		// NOTE: If the current session if in 'PINGING_SERVER' status, there is
+		// no need to reset the session. The destruction and creation of a new
+		// session in case this session has exceeded the time specified by
+		// 'connection_max_age_ms' will be deferred to the next time the session
+		// is used outside 'PINGING_SERVER' operation. For more context see #3502.
+		sess->status != PINGING_SERVER
 	) {
 		if (mysql_thread___reset_connection_algorithm == 2) {
 			sess->create_new_session_and_reset_connection(this);

